### ISSUE TITLE
Allow use of file descriptor instead of file name

### ISSUE THREAD

**Issue by horazont** [pyserial/pyserial#195]
This allows to use pyserial if you have already opened the serial port yourself.

This is for example handy if you are actually using a PTY instead of an actual serial port. Since PTYs are made to emulate TTYs, pyserial should work as-is. This helps a lot with creating mock applications for testing serial-based tools.

I may prepare a minimal pull request if desired.

**Comment by zsquareplusc** [comment]
This does not seem to be a feature that is usable across different platforms that pySerial supports.

Working with PTY is already possible, e.g. see this [test](https://github.com/pyserial/pyserial/blob/master/test/test_pty.py).  `os.ttyname()` can be used to translate the file handle into a device path.

**Comment by eras** [comment]
This works for in that scenario, however the output of this code:

    #!/usr/bin/env python3

    import os
    import pty

    (master, slave) = pty.openpty()
    print("master: {}".format(os.ttyname(master)))
    print("slave: {}".format(os.ttyname(slave)))

is

    master: /dev/ptmx
    slave: /dev/pts/59

and I'm pretty sure opening `/dev/ptmx` is not going to be helpful in this case.

**Comment by liberforce** [comment]
This would be useful if someone wants to connect to an external event loop like the GLib and [be notified when data arrives](https://stackoverflow.com/a/49387811/518853).

**Comment by chrismurf** [comment]
Went looking to see if this capability was supported, as I am using pty in exactly the way that @eras describes above.  Agree this would be a helpful addition to the library.

**Comment by TheSkorm** [comment]
I tested a work around which is replacing the fd that pyserial creates.
```
import serial
import pty
master, slave = pty.openpty()
ser = serial.Serial('/dev/ptmx')
ser.fd = master
ser.write("test")
```

**Comment by gagabla** [comment]
> ```
> master: /dev/ptmx
> slave: /dev/pts/59
> ```
> 
> and I'm pretty sure opening `/dev/ptmx` is not going to be helpful in this case.

I naively tried opening it and it doesn't work (but no error either). What I don't get is, why this is the case? I thought it is a symmetric pair of devices which can exchange data in both directions.
Since you already suspected it wouldn't work, would you be so kind to share your thoughts @eras ? Just because `/dev/ptmx` looks wrong, or is there some more fundamental difference between master and slave which I am missing?

**Comment by TheSkorm** [comment]
@gagabla that's because the `/dev/ptmx` device is a special control device. When you open up `/dev/ptmx` that file descriptor is paired with a specific `pts`. Opening `/dev/ptmx` again will trigger the creation of another `pts` device.

https://linux.die.net/man/4/ptmx

**Comment by gagabla** [comment]
Thanks @TheSkorm I always came across the pty manpage, but it is not as explicit as your link, so only know I understand the way this works.

It's a bit sad that no error is raised when using ttyname on the master handle and then passing this to pyserial, I think the better way would be to use ptsname instead of ttyname, but it is not exposed in python it seems.

But applying this new understanding to your workaround above, I get the impression it is not very clean because it will open an unnecessary master/slave pair and have it open till the process terminates?

**Comment by TheSkorm** [comment]
This is true, the impact has been minimal in my use cases that I haven't cared. Might be able to switch it out for /dev/null? or call close on the original fd.
